syntax = "proto3";

package PB;

message Node {
    string ip = 1;
    uint32 port = 2;
}

enum OpType {
    INVALID = 0;
    GET = 1;
    PUT = 2;
    DELETE = 3; // not used
}

// enum MessageType {
//     EMPTY = 0;
//     CLIENT_REQUEST = 1;
//     MERGE_REQUEST = 2;
// }

message Command {
    OpType type = 1;
    
    string key = 2;

    string value = 3;
}

message Txn {
    // not used;
    int64 txn_id = 1;
    
    repeated Command commands = 2;
}

message ClientRequest {

    Node dest_node = 1;

    Txn txn = 2;
}

message MergeRequest {
    
    Node src_node = 2;

    Node dest_node = 3;

    repeated Txn txns = 4;

}

message ClientReply {
    bool exec_res = 1;

    repeated string query_set = 2;
}